composer/callbacks/run_directory_uploader.py

```diff
@@ -0,0 +1,250 @@
+        files_to_be_uploaded = []
+        for root, dirs, files in os.walk(run_directory):
+            del dirs  # unused
+            for file in files:
+                filepath = os.path.join(root, file)
+                relpath = os.path.relpath(filepath, run_directory)  # chop off the run directory
+                modified_time = os.path.getmtime(filepath)
+                if modified_time > self._last_upload_timestamp:
+                    copied_path = os.path.join(self._upload_staging_folder, str(new_last_uploaded_timestamp), relpath)
+                    files_to_be_uploaded.append(relpath)
+                    copied_path_dirname = os.path.dirname(copied_path)
+                    os.makedirs(copied_path_dirname, exist_ok=True)
+                    # shutil.copyfile(filepath, copied_path)
+                    shutil.copy2(filepath, copied_path)
+                    self._file_upload_queue.put_nowait(copied_path)
+        self._last_upload_timestamp = new_last_uploaded_timestamp
+        # now log which files are being uploaded. OK to do, since we're done reading the directory,
+        # and any logfiles will now have their last modified timestamp
+        # incremented past self._last_upload_timestamp
+        logger.metric(log_level, {"run_directory/uploaded_files": files_to_be_uploaded})
+
+
+def _upload_worker(
+    file_queue: Union[queue.Queue[str], multiprocessing.JoinableQueue[str]],
+    is_finished: Union[multiprocessing._EventType, threading.Event],
+    upload_staging_dir: str,
+    provider_name: str,
+    container_name: str,
+    object_name_prefix: str,
+    init_kwargs: Dict[str, Any],
+):
+    """A long-running function to handle uploading files.
+
+    Args:
+        file_queue (queue.Queue or multiprocessing.JoinableQueue): The worker will poll
+            this queue for files to upload.
+        is_finished (threading.Event or multiprocessing.Event): An event that will be
+            set when training is finished and no new files will be added to the queue.
+            The worker will continue to upload existing files that are in the queue.
+            When the queue is empty, the worker will exit.
+        upload_staging_dir (str): The upload staging directory.
+        provider_name (str): The cloud provider name.
+        container_name (str): The container name (e.g. s3 bucket) for the provider
+            where files will be uploaded.
+        object_name_prefix (str): Prefix to prepend to the object names
+             before they are uploaded to the blob store.
+        init_kwargs (Dict[str, Any]): Arguments to pass in to the
+            :class:`~libcloud.storage.providers.Provider` constructor.
+    """
+    from libcloud.storage.providers import get_driver
+    provider_cls = get_driver(provider_name)
+    provider = provider_cls(**init_kwargs)
+    container = provider.get_container(container_name)
+    while True:
+        try:
+            file_path_to_upload = file_queue.get_nowait()
```

Comment:
Cleaner to do `file_queue.get(block=True, timeout=0.5)` to avoid having to do the `time.sleep()`